File: frontend/src/types/provider.ts
export interface Provider {
  id: number
  type: string
  category: 'llm' | 'tts'
  name: string
  base_url?: string
  api_key: string          // masked: "sk-***xxxx"
  enabled: boolean
  extra_config?: {
    cached_models?: string[]
    cached_at?: string
  }
  selected_models?: string[]
  created_at?: string
  updated_at?: string
}

export interface ProviderCreate {
  type: string
  category?: string
  name: string
  base_url?: string
  api_key: string
  enabled?: boolean
}

export interface ProviderUpdate {
  type?: string
  category?: string
  name?: string
  base_url?: string
  api_key?: string
  enabled?: boolean
  selected_models?: string[]
}

export interface ProviderTypeInfo {
  type: string
  category: 'llm' | 'tts'
  requires_base_url: boolean
  default_base_url: string | null
  supports_list_models: boolean
}

File: backend/models/provider.py
from datetime import datetime, timezone
from sqlalchemy import Column, Integer, String, Text, Boolean, JSON, DateTime

from backend.database import Base


class Provider(Base):
    __tablename__ = "providers"

    id = Column(Integer, primary_key=True)
    type = Column(String(32), nullable=False)
    name = Column(String(64), nullable=False, unique=True)
    base_url = Column(String(256), nullable=True)
    api_key_encrypted = Column(Text, nullable=False)
    enabled = Column(Boolean, default=True, nullable=False)
    category = Column(String(16), default="llm", nullable=False)
    extra_config = Column(JSON, default=dict)
    selected_models = Column(JSON, default=list)
    created_at = Column(DateTime, default=lambda: datetime.now(timezone.utc))
    updated_at = Column(
        DateTime,
        default=lambda: datetime.now(timezone.utc),
        onupdate=lambda: datetime.now(timezone.utc),
    )
